apps/prairielearn/src/pages/instructorInstanceAdminSettings/components/SelfEnrollmentSettings.tsx

```diff
@@ -86,6 +152,158 @@ export function SelfEnrollmentSettings({
+
+  const selfEnrollmentEnabled = watch('self_enrollment_enabled');
+  const selfEnrollmentRequiresSecretLink = watch('self_enrollment_requires_secret_link');
+  const selfEnrollmentEnabledBeforeDate = watch('self_enrollment_enabled_before_date');
+
+  const [showDateInput, setShowDateInput] = useState(!!selfEnrollmentEnabledBeforeDate);
+
+  return (
+    <>
+      <h2 class="h4">Self-enrollment</h2>
+
+      <div class="mb-3 form-check">
+        <input
+          class="form-check-input"
+          type="checkbox"
+          id="self_enrollment_enabled"
+          disabled={!canEdit || !enrollmentManagementEnabled}
+          {...register('self_enrollment_enabled')}
+          name="self_enrollment_enabled"
+        />
+        <label class="form-check-label" for="self_enrollment_enabled">
+          Enable self-enrollment
+        </label>
+        <div class="small text-muted">
+          Allow students to enroll themselves in this course instance.
+        </div>
+      </div>
+
+      <div class="mb-3 form-check">
+        <input
+          class="form-check-input"
+          type="checkbox"
+          id="hide_in_enroll_page"
+          disabled={!canEdit || !selfEnrollmentEnabled || selfEnrollmentRequiresSecretLink}
+          {...register('hide_in_enroll_page')}
+          name="hide_in_enroll_page"
+        />
+        <label class="form-check-label" for="hide_in_enroll_page">
+          Hide in enrollment page
+        </label>
+        <div class="small text-muted">
+          If enabled, hides the course instance in the enrollment page, so that only direct links to
+          the course can be used for enrollment.
+        </div>
+      </div>
+
+      <div class="mb-3 form-check">
+        <input
+          class="form-check-input"
+          type="checkbox"
+          id="self_enrollment_requires_secret_link"
+          disabled={!canEdit || !selfEnrollmentEnabled || !enrollmentManagementEnabled}
+          {...register('self_enrollment_requires_secret_link')}
+          name="self_enrollment_requires_secret_link"
+        />
+        <label class="form-check-label" for="self_enrollment_requires_secret_link">
+          Self-enrollment requires secret link
+        </label>
+        <div class="small text-muted">
+          If enabled, self-enrollment requires a secret link to enroll. If disabled, any link to the
+          course instance will allow self-enrollment.
+        </div>
+      </div>
+
+      <div class="mb-3 form-check">
+        <input
+          class="form-check-input"
+          type="checkbox"
+          id="disable_self_enrollment_after_date"
+          disabled={!canEdit || !selfEnrollmentEnabled || !enrollmentManagementEnabled}
+          checked={showDateInput}
+          onChange={(e) => {
+            const target = e.target as HTMLInputElement;
+            setShowDateInput(target.checked);
+
+            // Clear the date when unchecking
```

Comment:
We need to do this because we don't represent this checkbox in the JSON schema anywhere. I personally think this is fine.

Reply:
An alternative would be a solution that preserves it client-side, but on a page reload it is blown away.